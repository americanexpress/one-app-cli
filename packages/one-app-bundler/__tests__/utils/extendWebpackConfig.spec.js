/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const extendWebpackConfig = require('../../utils/extendWebpackConfig');
const getConfigOptions = require('../../utils/getConfigOptions');
const getCliOptions = require('../../utils/getCliOptions');

jest.mock('read-pkg-up', () => ({
  sync: jest.fn(() => ({ pkg: { name: 'test-module', version: '1.0.0' } })),
}));

jest.mock('/path/webpack.config.js', () => ({
  module: {
    rules: [
      { test: /\.js$/, use: 'my-super-cool-loader' },
    ],
  },
}), { virtual: true });

jest.mock('/path/webpack.client.config.js', () => ({
  module: {
    rules: [
      { test: /\.js$/, use: 'my-super-cool-client-loader' },
    ],
  },
}), { virtual: true });

jest.mock('/path/webpack.server.config.js', () => ({
  module: {
    rules: [
      { test: /\.js$/, use: 'my-super-cool-server-loader' },
    ],
  },
}), { virtual: true });

jest.mock('../../utils/getConfigOptions', () => jest.fn(() => ({})));
jest.mock('../../utils/getCliOptions', () => jest.fn(() => ({})));

describe('extendWebpackConfig', () => {
  let originalWebpackConfig = {};

  jest.spyOn(process, 'cwd').mockImplementation(() => '/path');

  const originalPlatform = process.platform;

  beforeEach(() => {
    Object.defineProperty(process, 'platform', { value: originalPlatform });
    jest.clearAllMocks();
    originalWebpackConfig = {
      resolve: {
        mainFields: ['browser', 'module', 'main'],
        modules: ['.', 'node_modules'],
        extensions: ['.js', '.jsx'],
      },
      module: {
        rules: [
          {
            test: /\.jsx?$/,
            include: ['src', 'node_modules'],
            use: ['babel-loader'],
          },
        ],
      },
    };
  });

  it('should apply webpack config', () => {
    getConfigOptions.mockReturnValueOnce({ webpackConfigPath: 'webpack.config.js' });

    const result = extendWebpackConfig(originalWebpackConfig);
    const { rules } = result.module;
    const lastRule = rules[rules.length - 1];
    expect(rules.length).toBe(originalWebpackConfig.module.rules.length + 1);
    expect(lastRule.use).toBe('my-super-cool-loader');
  });

  it('should apply a client webpack config', () => {
    getConfigOptions.mockReturnValueOnce({ webpackClientConfigPath: 'webpack.client.config.js' });
    const result = extendWebpackConfig(originalWebpackConfig, 'client');
    const { rules } = result.module;
    const lastRule = rules[rules.length - 1];
    expect(rules.length).toBe(originalWebpackConfig.module.rules.length + 1);
    expect(lastRule.use).toBe('my-super-cool-client-loader');
  });

  it('should apply a server webpack config', () => {
    getConfigOptions.mockReturnValueOnce({ webpackServerConfigPath: 'webpack.server.config.js' });
    const result = extendWebpackConfig(originalWebpackConfig, 'server');
    const { rules } = result.module;
    const lastRule = rules[rules.length - 1];
    expect(rules.length).toBe(originalWebpackConfig.module.rules.length + 1);
    expect(lastRule.use).toBe('my-super-cool-server-loader');
  });

  it('should bundle requiredExternals designated by providedExternals', () => {
    getConfigOptions.mockReturnValueOnce({ providedExternals: ['ajv', 'chalk', 'lodash'], moduleName: 'test-root-module' });
    const result = extendWebpackConfig(originalWebpackConfig);
    const { rules } = result.module;
    expect(rules).toHaveLength(originalWebpackConfig.module.rules.length + 1);
    expect(rules[rules.length - 1]).toMatchSnapshot();
  });

  it('should use the provided requiredExternals configured', () => {
    getConfigOptions.mockReturnValueOnce({ requiredExternals: ['ajv', 'lodash'] });
    const result = extendWebpackConfig(originalWebpackConfig);
    const { rules } = result.module;

    expect(rules).toHaveLength(originalWebpackConfig.module.rules.length + 3);
    expect(rules[rules.length - 3]).toMatchSnapshot({
      test: expect.stringMatching(/ajv\/$/),
    });
    expect(rules[rules.length - 2]).toMatchSnapshot({
      test: expect.stringMatching(/lodash\/$/),
    });
    expect(rules[rules.length - 1]).toMatchSnapshot();
  });

  it('should use the correct trailing slash on windows', () => {
    Object.defineProperty(process, 'platform', { value: 'win32' });
    getConfigOptions.mockReturnValueOnce({ requiredExternals: ['ajv', 'lodash'] });
    const result = extendWebpackConfig(originalWebpackConfig);
    const { rules } = result.module;

    expect(rules).toHaveLength(originalWebpackConfig.module.rules.length + 3);
    expect(rules[rules.length - 3]).toMatchSnapshot({
      test: expect.stringMatching(/ajv\\$/),
    });
    expect(rules[rules.length - 2]).toMatchSnapshot({
      test: expect.stringMatching(/lodash\\$/),
    });
    expect(rules[rules.length - 1]).toMatchSnapshot();
  });

  it('should validate the one app version', () => {
    getConfigOptions.mockReturnValueOnce({ appCompatibility: '^4.41.0' });
    const result = extendWebpackConfig(originalWebpackConfig);
    const { rules } = result.module;
    expect(rules[rules.length - 1]).toMatchSnapshot();
  });

  it('should enable watch mode', () => {
    getCliOptions.mockReturnValueOnce({ watch: true });
    const result = extendWebpackConfig(originalWebpackConfig);
    expect(result.watch).toBe(true);
    expect(result.watchOptions).toMatchObject({
      aggregateTimeout: 500,
      ignored: /node_modules/,
    });
  });
});
