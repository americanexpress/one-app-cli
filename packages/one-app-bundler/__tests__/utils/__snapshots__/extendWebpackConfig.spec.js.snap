// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extendWebpackConfig should bundle requiredExternals designated by providedExternals 1`] = `
Object {
  "test": "/path/src/index",
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/provided-externals-loader",
      "options": Object {
        "moduleName": "test-root-module",
        "providedExternals": Array [
          "ajv",
          "chalk",
          "lodash",
        ],
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the correct trailing slash on windows 1`] = `
Object {
  "test": StringMatching /ajv\\\\\\\\\\$/,
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/externals-loader",
      "options": Object {
        "externalName": "ajv",
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the correct trailing slash on windows 2`] = `
Object {
  "test": StringMatching /lodash\\\\\\\\\\$/,
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/externals-loader",
      "options": Object {
        "externalName": "lodash",
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the correct trailing slash on windows 3`] = `
Object {
  "test": "/path/src/index",
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/validate-required-externals-loader",
      "options": Object {
        "requiredExternals": Array [
          "ajv",
          "lodash",
        ],
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the provided requiredExternals configured 1`] = `
Object {
  "test": StringMatching /ajv\\\\/\\$/,
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/externals-loader",
      "options": Object {
        "externalName": "ajv",
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the provided requiredExternals configured 2`] = `
Object {
  "test": StringMatching /lodash\\\\/\\$/,
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/externals-loader",
      "options": Object {
        "externalName": "lodash",
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should use the provided requiredExternals configured 3`] = `
Object {
  "test": "/path/src/index",
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/validate-required-externals-loader",
      "options": Object {
        "requiredExternals": Array [
          "ajv",
          "lodash",
        ],
      },
    },
  ],
}
`;

exports[`extendWebpackConfig should validate the one app version 1`] = `
Object {
  "test": "/path/src/index",
  "use": Array [
    Object {
      "loader": "@americanexpress/one-app-bundler/webpack/loaders/validate-one-app-compatibility-loader",
      "options": Object {
        "appCompatibility": "^4.41.0",
      },
    },
  ],
}
`;
